### PR TITLE
[clang-cache] For include-tree disable `-gmodules` to avoid debug info referencing a non-existent PCH filename

### PR DESCRIPTION
**Pull request by akyrtzi**
rdar://104493507

### CODE REVIEW

**Comment by akyrtzi** [comment]
@swift-ci Please test

**Comment by akyrtzi** [comment]
@swift-ci Please test

**Review by benlangmuir — COMMENTED** [review]
Can we move this into `configureInvocationForCaching`? This shouldn't be specific to clang-cache.

**Comment by akyrtzi** [comment]
> Can we move this into `configureInvocationForCaching`? This shouldn't be specific to clang-cache.

I consider this a temporary workaround for `clang-cache`, we should properly support `-gmodules` for clang explicit modules support.

**Comment by benlangmuir** [comment]
> I consider this a temporary workaround for clang-cache, we should properly support -gmodules for clang explicit modules support.

The same issue that applies to clang-cache also affects command lines produced by clang-scan-deps already. If we add support for include-tree in libclang it will be an issue there as well _until_ we actually fix it.  If in the future we solve this issue for explicit modules but not for clang-cache, we could always move the code here later.

**Comment by cachemeifyoucan** [comment]
Does `-gmodule` work without `-fmodule`?

**Comment by akyrtzi** [comment]
> Does `-gmodule` work without `-fmodule`?

Yes, it works for the PCH irrespective of whether modules are used or not.

**Comment by akyrtzi** [comment]
@swift-ci Please test

**Comment by akyrtzi** [comment]
> The same issue that applies to clang-cache also affects command lines produced by clang-scan-deps already. If we add support for include-tree in libclang it will be an issue there as well _until_ we actually fix it. If in the future we solve this issue for explicit modules but not for clang-cache, we could always move the code here later.

Good point, I've moved the change to `configureInvocationForCaching`.

File: clang/lib/Tooling/DependencyScanning/ScanAndUpdateArgs.cpp [hunk 1]
```diff
@@ -54,6 +54,9 @@ void tooling::dependencies::configureInvocationForCaching(
       PPOpts.MacroIncludes.clear();
       PPOpts.Includes.clear();
     }
+    // Disable `-gmodules` to avoid debug info referencing a non-existent PCH
+    // filename.
+    CI.getCodeGenOpts().DebugTypeExtRefs = false;
```
Comment (benlangmuir) [inline clang/lib/Tooling/DependencyScanning/ScanAndUpdateArgs.cpp:62]: Do you think we should also reset `getHeaderSearchOpts().ModuleFormat` to "raw"? Technically it probably doesn't matter, but it's consistent with -go-modules which controls both options.

**Review by benlangmuir — APPROVED** [review]
Thanks. LGTM, though left a question about ModuleFormat.

**Comment by akyrtzi** [comment]
@swift-ci Please test

File: clang/lib/Tooling/DependencyScanning/ScanAndUpdateArgs.cpp (hunk 1 above)
Comment (akyrtzi) [reply clang/lib/Tooling/DependencyScanning/ScanAndUpdateArgs.cpp:62]: Code already resets `HeaderSearchOptions`; I've added a note that we'd have to preserve `HeaderSearchOptions::ModuleFormat` when properly supporting `-gmodules`.

**Pull request merged**